### PR TITLE
[dbnode] Avoid resultsMap memory leak during query execution

### PR DESCRIPTION
**Pull request by linasm**
**What this PR does / why we need it**:
In environments with no (or very low) churn from `QueryResultsPool` (this can be checked with `index_query_results_pool_get_on_empty` metrics), pooled `results` objects would never get evicted, and the `ResultsMap` that they contain would grow to fit the largest query result ever observed. Over time, this would potentially build up to gigabytes of memory.
This PR introduces a change where "too large" objects would be discarded from pooling. It is difficult to say exactly what is "too large" (as this depends on the environment), so a heuristic is used to make a decision - when the "capacity" of the map is underutilized for a given number of times in a row, the object gets discarded.

Before the change (3.3 GB and growing):
Screenshot 2022-01-04 at 11 06 22

After the change (a few MB, stable):
Screenshot 2022-01-04 at 11 09 26

Total memory usage comparison (bottom line on the graph is a build containing this change):
Screenshot 2022-01-04 at 11 10 45

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
NONE

**Does this PR require updating code package or user-facing documentation?**:
NONE

### CODE REVIEW

File: src/dbnode/storage/index/results.go [hunk 1]
```diff
@@ -48,6 +48,10 @@ type results struct {
 	resultsMap     *ResultsMap
 	totalDocsCount int
 
+	// Utilization stats, do not reset:
+	maxObservedSize        int
```
Comment (rallen090) [inline src/dbnode/storage/index/results.go:52]: I'd probably name this `resultsMapCapacity` - too bad there is no `cap(map)`...

Comment (rallen090) [reply src/dbnode/storage/index/results.go:52]: And maybe the other one `consecutiveTimesUnderCapacity`

File: src/dbnode/storage/index/results.go [hunk 2]
```diff
@@ -178,11 +182,27 @@ func (r *results) TotalDocsCount() int {
...
+	if r.totalDocsCount < r.maxObservedSize {
+		r.underutilizedRunLength++
+	} else {
+		r.maxObservedSize = r.totalDocsCount
+		r.underutilizedRunLength = 0
+	}
+
+	// If the size of reusable result object is not fully utilized for a long enough time,
+	// we want to drop it from the pool and prevent holding on to excessive memory indefinitely.
+	putBackToPool := r.underutilizedRunLength < 100
```
Comment (rallen090) [inline src/dbnode/storage/index/results.go:196]: Nice I like this approach, where we should purge pooled results sizes unless they are common

File: src/dbnode/storage/index/results.go (hunk 1 above)
Comment (linasm) [reply src/dbnode/storage/index/results.go]: 👍

**Review by rallen090 — APPROVED** [review]

**Pull request merged**